geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java

```diff
@@ -3803,6 +3804,14 @@ private void setMemoryThresholdReachedCounterTrue(final DistributedMember idm) {
     }
   }
 
+  @Override
+  protected MemoryThresholdInfo getAtomicThresholdInfo() {
+    synchronized (this.memoryThresholdReachedMembers) {
```

Comment:
I think this change could impact performance of writes. So I would keep the initial check of the atomic outside the sync. Only if the atomic is true then go inside the sync just as you do now. Once you get in their you may find the atomic is now false and that is okay. Most of the time this method is called the atomic is false and in that case we do not want it doing synchronization nor object allocation.
Many ways exist of doing this but I suggest that in the case when it is false your return a canonical that always returns false and an empty set. You could create a static final instance of this (to prevent object allocation when the atomic is false) like so: "new MemoryThresholdInfo(false, Collections.emptySet<>())".
So then this method would be:
if (!this.memoryThresholdReached.get()) {
  return MemoryThresholdInfo.NOT_REACHED; // or whatever name you come up for your static final
}
after this would be your existing code that creates a real MemoryThresholdInfo.